server/src/main/java/org/opensearch/index/shard/IndexShard.java

```diff
@@ -4531,19 +4523,58 @@ public void syncSegmentsFromRemoteSegmentStore(boolean overrideLocal, boolean re
                     skippedSegments.add(file);
                 }
             }
+
             if (refreshLevelSegmentSync && segmentInfosSnapshotFilename != null) {
                 try (
                     ChecksumIndexInput indexInput = new BufferedChecksumIndexInput(
                         storeDirectory.openInput(segmentInfosSnapshotFilename, IOContext.DEFAULT)
                     )
                 ) {
-                    SegmentInfos infosSnapshot = SegmentInfos.readCommit(
-                        store.directory(),
-                        indexInput,
-                        Long.parseLong(segmentInfosSnapshotFilename.split("__")[1])
-                    );
+                    SegmentInfos infosSnapshot = null;
+                    boolean canRetry = true;
+                    while (true) {
+                        try {
+                            infosSnapshot = SegmentInfos.readCommit(
+                                store.directory(),
+                                indexInput,
+                                Long.parseLong(segmentInfosSnapshotFilename.split("__")[1])
+                            );
+                            break;
+                        } catch (FileNotFoundException e) {
+                            /**
+                             * While we're downloading the segment files from remote store, primary is continuously writing as well.
+                             * Primary updates segmentInfo in same file on remote store during refreshes(not for commit though).
+                             * This can lead to race conditions where the segmentInfo file downloaded is from a newer refresh
+                             * and points to segment files which we haven't even downloaded thus resulting in FileNotFoundException.
+                             * So to handle this, we again download the delta in segments from store and retry.
+                             * Since, we're not downloading the files which are already downloaded, we skip the segmentInfosSnapshot during retry.
+                             * This guarantees that all files needs to refresh the reader are available, and we don't run into race conditions during retry.
+                             */
+                            if (canRetry) {
```

Comment:
Thanks for the detailed comment on what this change is trying to do. It makes sense and should work as intended.

But the root cause of this problem is not having 1:1 mapping between metadata and snapshot file. I am thinking of combining these 2 files and provide SegmentInfos as part of metadata file. Let me know what you think.

Reply:
Sure, combining metadata and segment info should work out. Not sure if there are any downsides with introducing that coupling though.

Reply:
Am also ok with this as a separate issue that we can complete before GA.

Reply:
Tracking issue: https://github.com/opensearch-project/OpenSearch/issues/7806